pkg/reconciler/apis/apiexportendpointslice/apiexportendpointslice_reconcile.go

```diff
@@ -0,0 +1,174 @@
+	"context"
+	"fmt"
+	"net/url"
+	"path"
+
+	"github.com/kcp-dev/logicalcluster/v2"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+
+	virtualworkspacesoptions "github.com/kcp-dev/kcp/cmd/virtual-workspaces/options"
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	conditionsv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/apis/conditions/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/util/conditions"
+	"github.com/kcp-dev/kcp/pkg/logging"
+	apiexportbuilder "github.com/kcp-dev/kcp/pkg/virtual/apiexport/builder"
+)
+
+func (c *controller) reconcile(ctx context.Context, apiExportEndpointSlice *apisv1alpha1.APIExportEndpointSlice) error {
+
+	// TODO (fgiloux): When the information is available in the cache server
+	// check if at least one APIBinding is bound in the shard to the APIExport referenced by the APIExportEndpointSLice.
+	// If so, add the respective endpoint to the status.
+	// For now the unfiltered list is added.
+
+	// Check for valid reference
+	workspaceRef := apiExportEndpointSlice.Spec.APIExport.Workspace
+	if workspaceRef == nil {
+		conditions.MarkFalse(
+			apiExportEndpointSlice,
+			apisv1alpha1.APIExportValid,
+			apisv1alpha1.APIExportInvalidReferenceReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			"Missing APIExport reference",
+		)
+		return nil
+	}
+
+	// Get APIExport
+	apiExportClusterName, err := getAPIExportClusterName(apiExportEndpointSlice)
+	if err != nil {
+		conditions.MarkFalse(
+			apiExportEndpointSlice,
+			apisv1alpha1.APIExportValid,
+			apisv1alpha1.APIExportInvalidReferenceReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			err.Error(),
+		)
+		return nil
+	}
+
+	apiExport, err := c.getAPIExport(apiExportClusterName, workspaceRef.ExportName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Don't keep the endpoints if the APIExport has been deleted
+			apiExportEndpointSlice.Status.APIExportEndpoints = nil
+			conditions.MarkFalse(
+				apiExportEndpointSlice,
+				apisv1alpha1.APIExportValid,
+				apisv1alpha1.APIExportNotFoundReason,
+				conditionsv1alpha1.ConditionSeverityError,
+				"APIExport %s|%s not found",
+				apiExportClusterName,
+				workspaceRef.ExportName,
+			)
+			return nil
+		} else {
+			conditions.MarkFalse(
+				apiExportEndpointSlice,
+				apisv1alpha1.APIExportValid,
+				apisv1alpha1.InternalErrorReason,
+				conditionsv1alpha1.ConditionSeverityError,
+				"Error getting APIExport %s|%s: %v",
+				apiExportClusterName,
+				workspaceRef.ExportName,
+				err,
+			)
+			return err
+		}
+	}
+	conditions.MarkTrue(apiExportEndpointSlice, apisv1alpha1.APIExportValid)
+
+	if err = c.updateEndpoints(ctx, apiExportEndpointSlice, apiExport); err != nil {
+		conditions.MarkFalse(
+			apiExportEndpointSlice,
+			apisv1alpha1.APIExportEndpointSliceURLsReady,
+			apisv1alpha1.ErrorGeneratingURLsReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			err.Error(),
```

Comment:
Similarly here, what is a user to do with an error to List?

Reply:
Same as above, even if there is no user recourse, they need to be aware of any issues. Some information is better than no information.